### PR TITLE
Fixed #34267 -- Fixed sliced QuerySet.union() crash.

### PR DESCRIPTION
**Pull request by panicofr**
Fixed a bug due to application of slicing on its parts as optimization when they was not alredy sliced.
Thanks Raphaël Stefanini for the report.

### CODE REVIEW

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -532,15 +532,12 @@ def get_combinator_sql(self, combinator, all):
...
         elif self.query.is_sliced and combinator == "union":
-            limit = (self.query.low_mark, self.query.high_mark)
             for compiler in compilers:
                 # A sliced union cannot have its parts elided as some of them
                 # might be sliced as well and in the event where only a single
                 # part produces a non-empty resultset it might be impossible to
                 # generate valid SQL.
                 compiler.elide_empty = False
-                if not compiler.query.is_sliced:
-                    compiler.query.set_limits(*limit)
```
Comment (felixxm) [inline django/db/models/sql/compiler.py:543]: Unfortunately, it causes `queries.test_qs_combinators.QuerySetSetOperationTests.test_union_with_select_related_and_first` crash on Oracle. It seems that Oracle requires `ORDER BY` and `LIMIT/OFFSET` in all compound queries when they are wrapped in a subquery. As far as I'm aware we can amend `test_union_with_select_related_and_first` as ordering is not crucial there:
```diff
diff --git a/tests/queries/test_qs_combinators.py b/tests/queries/test_qs_combinators.py
index 6358e537ae..1ed5b46241 100644
--- a/tests/queries/test_qs_combinators.py
+++ b/tests/queries/test_qs_combinators.py
@@ -314,15 +314,14 @@ class QuerySetSetOperationTests(TestCase):
         qs2 = base_qs.filter(name="a2")
         self.assertSequenceEqual(qs1.union(qs2).order_by("pk"), [a1, a2])
 
-    @skipUnlessDBFeature("supports_slicing_ordering_in_compound")
     def test_union_with_select_related_and_first(self):
         e1 = ExtraInfo.objects.create(value=7, info="e1")
         a1 = Author.objects.create(name="a1", num=1, extra=e1)
         Author.objects.create(name="a2", num=3, extra=e1)
-        base_qs = Author.objects.select_related("extra")
+        base_qs = Author.objects.select_related("extra").order_by()
         qs1 = base_qs.filter(name="a1")
         qs2 = base_qs.filter(name="a2")
-        self.assertEqual(qs1.union(qs2).first(), a1)
+        self.assertEqual(qs1.union(qs2).order_by("name").first(), a1)
 
     def test_union_with_first(self):
         e1 = ExtraInfo.objects.create(value=7, info="e1")
```

**Review by felixxm — COMMENTED** [review]
@panicofr Thanks :+1: I left comments.

File: tests/queries/test_qs_combinators.py [hunk 2]
```diff
@@ -636,3 +635,21 @@ def test_operator_on_combined_qs_error(self):
                         operator_func(qs, combined_qs)
                     with self.assertRaisesMessage(TypeError, msg % operator_):
                         operator_func(combined_qs, qs)
+
+    @skipUnlessDBFeature(
+        "supports_slicing_ordering_in_compound", "can_distinct_on_fields"
+    )
+    def test_union_with_cached_results(self):
```
Comment (felixxm) [inline tests/queries/test_qs_combinators.py:642]: To sum up, I'd add:
```python
    def test_union_slice_index(self):
        Celebrity.objects.create(name="Famous")
        c1 = Celebrity.objects.create(name="Very famous")

        qs1 = Celebrity.objects.filter(name__startswith="F")
        qs2 = Celebrity.objects.filter(name__startswith="Very")
        combined_qs = qs1.union(qs2).order_by("name")
        self.assertEqual(combined_qs[1], c1)
```

**Comment by panicofr** [comment]
@felixxm thanks, I will work on it asap 👍

File: tests/queries/test_qs_combinators.py [hunk 3]
```diff
@@ -636,3 +635,21 @@ def test_operator_on_combined_qs_error(self):
...
+        qs1 = Report.objects.all()
+        qs2 = qs1.filter(creator__isnull=True)
+        qs3 = (
+            qs1.filter(creator__isnull=False)
+            .order_by("creator_id")
+            .distinct("creator_id")
+        )
+        combined_qs = qs2.union(qs3)
+        combined_qs.count()
+        combined_qs[1]
```
[the remark this answers is not in the document]
Comment (felixxm) [reply tests/queries/test_qs_combinators.py:655]: Right, we need to modify it a bit, this should do the trick:
```python
    def test_union_slice_index(self):
        Celebrity.objects.create(name="Famous")
        Celebrity.objects.create(name="Famous")
        Celebrity.objects.create(name="Very famous")
        
        qs1 = Celebrity.objects.filter(name="nonexistent").values("name")
        qs2 = Celebrity.objects.values("name").distinct()
        combined_qs = qs1.union(qs2).order_by("name")
        self.assertEqual(combined_qs[1], {"name": "Very famous"})
```

[the remark this answers is not in the document]
Comment (felixxm) [reply tests/queries/test_qs_combinators.py:655]: Good, we can simplify it a bit:
```python
    def test_union_slice_index(self):
        Celebrity.objects.create(name="Famous")
        c1 = Celebrity.objects.create(name="Very famous")
        
        qs1 = Celebrity.objects.filter(name="nonexistent")
        qs2 = Celebrity.objects.all()
        combined_qs = qs1.union(qs2).order_by("name")
        self.assertEqual(combined_qs[1], c1)
```

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
@panicofr Thanks for updates :+1:

**Pull request merged**